src/recon_test/test_data_processor_projectors.cxx

```diff
@@ -0,0 +1,279 @@
+#include "stir/recon_buildblock/ProjMatrixByBinUsingRayTracing.h"
+#include "stir/recon_buildblock/ForwardProjectorByBinUsingProjMatrixByBin.h"
+#include "stir/recon_buildblock/BackProjectorByBinUsingProjMatrixByBin.h"
+#include "stir/ProjDataInMemory.h"
+#include "stir/SeparableCartesianMetzImageFilter.h"
+
+START_NAMESPACE_STIR
+
+/*!
+  \ingroup test
+  \brief Test class for GPU projectors
+*/
+class TestDataProcessorProjectors : public RunTests
+{
+public:
+    //! Constructor that can take some input data to run the test with
+    TestDataProcessorProjectors(const std::string &sinogram_filename, const float fwhm);
+
+    virtual ~TestDataProcessorProjectors() {}
+
+    void run_tests();
+protected:
+    std::string _sinogram_filename;
+    float _fwhm;
+    shared_ptr<ProjDataInMemory> _input_sino_sptr;
+    shared_ptr<VoxelsOnCartesianGrid<float> > post_data_processor_bck_proj();
+    void pre_data_processor_fwd_proj(const VoxelsOnCartesianGrid<float> &input_image);
+};
+
+TestDataProcessorProjectors::TestDataProcessorProjectors(const std::string &sinogram_filename, const float fwhm) :
+    _sinogram_filename(sinogram_filename),
+    _fwhm(fwhm)
+{
+}
+
+void
+TestDataProcessorProjectors::
+run_tests()
+{
+    try {
+        // Open sinogram
+        _input_sino_sptr = MAKE_SHARED<ProjDataInMemory>(*ProjData::read_from_file(_sinogram_filename));
+
+        // Back project
+        std::cerr << "Tests for post-data-processor back projection\n";
+        shared_ptr<VoxelsOnCartesianGrid<float> > bck_projected_im_sptr =
+                this->post_data_processor_bck_proj();
+
+        // Forward project
+        std::cerr << "Tests for pre-data-processor forward projection\n";
+        this->pre_data_processor_fwd_proj(*bck_projected_im_sptr);
+    }
+    catch(const std::exception &error) {
+        std::cerr << "\nHere's the error:\n\t" << error.what() << "\n\n";
+        everything_ok = false;
+    }
+    catch(...) {
+        everything_ok = false;
+    }
+}
+
+void
+get_data_processor(shared_ptr<SeparableCartesianMetzImageFilter<float> > &data_processor_sptr, const float fwhm)
+{
+    data_processor_sptr.reset(new SeparableCartesianMetzImageFilter<float>);
+    std::string buffer;
+    std::stringstream parameterstream(buffer);
+
+    parameterstream << "Separable Cartesian Metz Filter Parameters :=\n"
+                    << "x-dir filter FWHM (in mm):= " << fwhm << "\n"
+                    << "y-dir filter FWHM (in mm):= " << fwhm << "\n"
+                    << "z-dir filter FWHM (in mm):= " << fwhm << "\n"
+                    << "x-dir filter Metz power:= .0\n"
+                    << "y-dir filter Metz power:= .0\n"
+                    << "z-dir filter Metz power:=.0\n"
+                    << "END Separable Cartesian Metz Filter Parameters :=\n";
+    data_processor_sptr->parse(parameterstream);
+}
+
+Succeeded compare_images(const VoxelsOnCartesianGrid<float> &im_1, const VoxelsOnCartesianGrid<float> &im_2)
```

Comment:
I'd make these `DiscretisedDensity<3,float>` and make the method explicitly `static` (i.e. local to this file)